contracts/contracts/Tally.sol

```diff
@@ -0,0 +1,248 @@
+            .treeDepths();
+
+        (VkRegistry vkRegistry, IMACI maci, , ) = _poll.extContracts();
+
+        // Get the verifying key
+        VerifyingKey memory vk = vkRegistry.getTallyVk(
+            maci.stateTreeDepth(),
+            intStateTreeDepth,
+            voteOptionTreeDepth
+        );
+
+        // Get the public inputs
+        uint256 publicInputHash = genTallyVotesPublicInputHash(
+            _numSignUps,
+            _batchStartIndex,
+            _tallyBatchSize,
+            _newTallyCommitment
+        );
+
+        // Verify the proof
+        return verifier.verify(_proof, vk, publicInputHash);
+    }
+
+    function computeMerkleRootFromPath(
+        uint8 _depth,
+        uint256 _index,
+        uint256 _leaf,
+        uint256[][] memory _pathElements
+    ) internal pure returns (uint256) {
+        uint256 pos = _index % LEAVES_PER_NODE;
+        uint256 current = _leaf;
+        uint8 k;
+
+        uint256[LEAVES_PER_NODE] memory level;
+
+        for (uint8 i = 0; i < _depth; ++i) {
+            for (uint8 j = 0; j < LEAVES_PER_NODE; ++j) {
+                if (j == pos) {
+                    level[j] = current;
+                } else {
+                    if (j > pos) {
+                        k = j - 1;
+                    } else {
+                        k = j;
+                    }
+                    level[j] = _pathElements[i][k];
+                }
+            }
+
+            _index /= LEAVES_PER_NODE;
+            pos = _index % LEAVES_PER_NODE;
+            current = hash5(level);
+        }
+        return current;
+    }
+
+
+     /*
+     * @notice Verify the result generated of the tally.json
+     * @param _voteOptionIndex the index of the vote option to verify the correctness of the tally
+     * @param _tallyResult Flattened array of the tally
+     * @param _tallyResultProof Corresponding proof of the tally result
+     * @param _tallyResultSalt the respective salt in the results object in the tally.json
+     * @param _spentVoiceCreditsHash hashLeftRight(number of spent voice credits, spent salt)
+     * @param _perVOSpentVoiceCreditsHash hashLeftRight(merkle root of the no spent voice credits per vote option, perVOSpentVoiceCredits salt)
+     * @param _tallyCommitment newTallyCommitment field in the tally.json
+     * @return valid a boolean representing successful verification
+     */
+    function verifyTallyResult(
+        Poll _poll,
+        uint256 _voteOptionIndex,
+        uint256 _tallyResult,
+        uint256[][] memory _tallyResultProof,
+        uint256 _spentVoiceCreditsHash,
+        uint256 _perVOSpentVoiceCreditsHash,
+        uint256 _tallyCommitment
+    ) public view returns (bool) {
+        (, , , uint8 voteOptionTreeDepth) = _poll.treeDepths();
+        uint256 computedRoot = computeMerkleRootFromPath(
+            voteOptionTreeDepth,
+            _voteOptionIndex,
+            _tallyResult,
+            _tallyResultProof
```

Comment:
Is this correct? passing the proof here?

Reply:
haven't checked or test this function, it was never used.

Reply:
we might delete it if we don't need it

Reply:
Removed. It's never been called